File: C/Interpolation/Chebyshev/chebyshev.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Define M_PI if it's not defined in math.h (standard in many C environments)
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/**
 * Function for Chebyshev Interpolation (Node Generation)
 * 
 * @param func  - Function pointer to the function to evaluate
 * @param n     - Number of Chebyshev points
 * @param a     - Start of interval
 * @param b     - End of interval
 * @param x_out - Pre-allocated array to store Chebyshev nodes
 * @param y_out - Pre-allocated array to store function values at nodes
 */
void chebyshev(double (*func)(double), int n, double a, double b, double *x_out, double *y_out) {
    for (int i = 0; i < n; i++) {
        // The index 'k' in the formula (2k-1) corresponds to (i+1) where i starts at 0
        // Formula: cos( (2k-1)*pi / (2n) )
        double k = i + 1.0;
        double cos_term = cos((2.0 * k - 1.0) * M_PI / (2.0 * n));
        
        // Map nodes from [-1, 1] to [a, b]
        x_out[i] = 0.5 * ((b - a) * cos_term + (b + a));
        
        // Evaluate function at the computed node
        y_out[i] = func(x_out[i]);
    }
}

// --- Test Case: Runge's Function ---
// f(x) = 1 / (1 + x^2)
double runge_function(double x) {
    return 1.0 / (1.0 + x * x);
}

int main() {
    int n = 10;
    double a = -5.0;
    double b = 5.0;

    // Allocate memory for nodes and values
    double *x_cheb = (double *)malloc(n * sizeof(double));
    double *y_cheb = (double *)malloc(n * sizeof(double));

    // Generate Chebyshev points
    chebyshev(runge_function, n, a, b, x_cheb, y_cheb);

    // Display Results
    printf("Chebyshev nodes and function values (n=%d, [%.1f, %.1f]):\n", n, a, b);
    printf("%-15s %-15s\n", "x_cheb", "y_cheb");
    printf("------------------------------------\n");
    for (int i = 0; i < n; i++) {
        printf("%-15.6f %-15.6f\n", x_cheb[i], y_cheb[i]);
    }

    // Free memory
    free(x_cheb);
    free(y_cheb);

    return 0;
}

File: Python/Interpolation/Chebyshev/chebyshev.py
import numpy as np

def chebyshev(func, n, a, b):
    """
    Function for Chebyshev Interpolation
    
    Parameters:
    func (callable): Function to interpolate (e.g., lambda x: ...)
    n (int): Number of Chebyshev points
    a (float): Start of interval
    b (float): End of interval
    
    Returns:
    x_cheb (numpy.ndarray): Chebyshev nodes mapped to [a, b]
    y_cheb (numpy.ndarray): Function values at Chebyshev nodes
    """
    
    # Generate index array k = 1, 2, ..., n
    k = np.arange(1, n + 1)
    
    # Calculate Chebyshev nodes on [-1, 1] and map them to [a, b]
    # Formula: 0.5 * ((b - a) * cos((2k - 1) * pi / (2n)) + (b + a))
    x_cheb = 0.5 * ((b - a) * np.cos((2 * k - 1) * np.pi / (2 * n)) + (b + a))
    
    # Evaluate the function at these nodes
    y_cheb = func(x_cheb)
    
    return x_cheb, y_cheb

# --- Test Driver ---
if __name__ == "__main__":
    # Define Runge's function: f(x) = 1 / (1 + x^2)
    func = lambda x: 1.0 / (1.0 + x**2)
    
    a = -5
    b = 5
    n = 10
    
    # Call the function
    x_nodes, y_values = chebyshev(func, n, a, b)
    
    print("Chebyshev nodes and function values:")
    print(f"{'x':>12} | {'y':>12}")
    print("-" * 27)
    
    # Print results in two columns
    for x_val, y_val in zip(x_nodes, y_values):
        print(f"{x_val:12.6f} | {y_val:12.6f}")